### ISSUE TITLE
Prevent DOS from HTTP Auth dialogs

### ISSUE THREAD

**Issue by wolfbeast** [MoonchildProductions/UXP#951]
C/P from BZ for brevity, with edits.

**The issue**

The basic auth prompt is window modal and blocks the entire browser from being interacted with when open. This can be easily abused by websites to effectively prevent users from using the browser ("DOS"). It is made extra bad by the fact that the dialog shows a credentials field and that websites can provide a message to the user. This is often combined with locking the user in DOM fullscreen, auto-playing sound and offering extension installs to "fix" some "virus".

This bug has gone unsolved for a very long time because a previous attempt at short term mitigation failed and other possible solutions pose a significant engineering challenge that no experienced engineers have been available to work on.

**Proposed short term mitigation**

The fix in [bug 1312243](https://bugzilla.mozilla.org/show_bug.cgi?id=1312243) introduced a maximum of three consecutive cancellations (controlled by a pref) that a user could perform until the browser would prevent the page from showing more dialogs.

This, in my opinion, is a great idea. The implementation, however, has a major fallacy: It checks the inner window id in the well-meaning attempt to find user navigation or reloads and clears its internal counter when that window id changes. Unfortunately this also clears the counter on non-user-initiated navigations and reloads. I believe that the true intention of the patch was to cancel the auth dialog after 3 attempts, except if:

- The user reloads the page on their own terms
- The user navigates to a different site on their own

Which is what I plan to implement, using the same pattern we applied to implement temporarily blocked site permissions:

- Temporarily store basic auth counter state on the browser object, as a map from baseDomain (eTLD+1) to number of cancellations
- Reset this state only on user initiated reload
- Reset the counter for a domain if the user has entered login data into the dialog and submitted

This would mitigate the DOS issue while hopefully not breaking any sites that rely on basic auth.

~~**Long term solution**~~

~~While the above proposal does help against a DOS attack, the simple fact that a website can spawn an application modal dialog that forces you to enter your credentials is, to say the least, not optimal. To properly and finally solve the full extent of this issue, I think they main BZ bug to tackle is:~~
- ~~Bug 613785, which makes the auth dialog tab modal.~~

TL;DR:
Short term solution: limit number of prompts. https://bugzilla.mozilla.org/show_bug.cgi?id=377496
~~Long term solution: make http auth dialog tab-modal https://bugzilla.mozilla.org/show_bug.cgi?id=613785~~

**Comment by mattatobin** [comment]
I am against tab model http auth because it is browser centric. I have no way to know how this or the other solution will affect non-browser applications.

You will need to preserve the orginal behavor by pref.

**Comment by mattatobin** [comment]
The question is more of using a browser element that doesn't come from tabbrowser or doing it purely from javascript.

**Comment by wolfbeast** [comment]
Fair enough.

**Comment by mattatobin** [comment]
I mean if you WANT to go non-model like JS alerts when spawned from INSIDE a browser element in tabbrowser then fair enough but you will have to preserve the original code path for other cases.

**Comment by wolfbeast** [comment]
Honestly I just want to prevent DOS effectively. Tab-modal would also solve the problem of domain confusion if 2 sites throw up auth prompts, but I see the potential issues with it being tab-modal as well.

**Comment by mattatobin** [comment]
I don't think domain confusion is a problem because that info is stated in the dialog.

----

![67275001-7d14-4e11-a64a-af2239d9f7fd](https://user-images.githubusercontent.com/3889475/51970889-fb5c8a00-2445-11e9-8670-04a75d3a39dc.jpg)

![90d40261-edcc-48ce-88f0-4808c414f8f1](https://user-images.githubusercontent.com/3889475/51970890-fb5c8a00-2445-11e9-91f3-19a56005c437.jpg)
